astropy/coordinates/spectral_coordinate.py

```diff
@@ -0,0 +1,842 @@
+        # The vectors are aligned but may be in opposite directions, so we use
+        # the dot product to determine this.
+
+        diff_vel = fin_vel - init_proj_vel
+        delta_vel = np.dot(diff_vel, line_of_sight_unit_vec)
+
+        if NUMPY_LT_1_17:
+            delta_vel *= diff_vel.unit
+
+        # In the case where the projected velocity is nan, we can assume that
+        # the final velocity different is zero, and thus the actual velocity
+        # delta is equal to the original radial velocity.
+
+        # TODO: Due to lack of precision in some coordinate transformations,
+        # we may end up with a final velocity very close to, but not quite at,
+        # zero. In this case, set a tolerance for the final velocity; if it's
+        # below this tolerance, assume the delta velocity is essentially nan.
+
+        fin_vel_mag = np.linalg.norm(fin_vel)
+
+        if NUMPY_LT_1_17:
+            fin_vel_mag *= fin_vel.unit
+
+        if np.isnan(delta_vel) or fin_vel_mag < 1e-7 * fin_vel.unit:
+            delta_vel = -self.radial_velocity
+
+        return self._apply_relativistic_doppler_shift(delta_vel)
+
+    def with_observer_velocity(self, frame, preserve_observer_frame=False):
+        """
+        Alters the velocity of the observer, but not the position.
+
+        If a coordinate frame is specified, the observer velocities will be
+        modified to be stationary in the specified frame. If a coordinate
+        instance is specified, optionally with non-zero velocities, the
+        observer velocities will be updated so that the observer is co-moving
+        with the specified coordinates.
+
+        Parameters
+        ----------
+        frame : `~astropy.coordinates.BaseCoordinateFrame` or `~astropy.coordinates.SkyCoord`
+            The observation frame containing the new velocity for the observer.
+        preserve_observer_frame : bool
+            If `True`, the final observer frame class will be the same as the
+            original one, and if `False` it will be the frame of the velocity
+            reference class.
+
+        Returns
+        -------
+        new_coord : `SpectralCoord`
+            The new coordinate object representing the spectral data
+            transformed based on the observer's new velocity frame.
+        """
+
+        if self.observer is None or self.target is None:
+            raise ValueError("This method can only be used if both observer "
+                             "and target are defined on the SpectralCoord.")
+
+        if hasattr(frame, 'frame'):
+            frame = frame.frame
+        elif isinstance(frame, type):
+            frame_cls = frame
+            frame = frame_cls(0 * u.m, 0 * u.m, 0 * u.m,
+                              0 * u.m / u.s, 0 * u.m / u.s, 0 * u.m / u.s,
+                              representation_type='cartesian',
+                              differential_type='cartesian')
+        elif isinstance(frame, str):
```

Comment:
Maybe swap branches so that the initialization with `0` only  has to be done once?
```
if hasattr...
    ...
else:
    if instance(frame, str):
        ...
    if isinstance(frame, type):
        ...
```

Reply:
Done